tests/test_input_outputs.py

```diff
@@ -251,3 +254,71 @@ def test_BaseWrapper_postprocess_y():
     np.testing.assert_equal(y_postprocessed, y_array)
     extra_args = BaseWrapper.postprocess_y(y_array)[1]
     assert len(extra_args) == 0
+
+
+@pytest.mark.parametrize(
+    "X_dtype", ["float32", "float64", "int64", "int32", "uint8", "uint16"]
+)
+@pytest.mark.parametrize("y_dtype", ["int64", "int32", "uint8", "uint16"])
+@pytest.mark.parametrize("s_w_dtype", ["float32", "float64"])
+@pytest.mark.parametrize("run_eagerly", [True, False])
+def test_classifier_handles_types(X_dtype, y_dtype, s_w_dtype, run_eagerly):
+    """Tests that classifiers correctly handle dtype conversions and
+    return the same dtype as the inputs.
+    """
+    n, d = 100, 20
+    n_classes = 10
+    X = np.random.uniform(size=(n, d)).astype(X_dtype)
+    y = np.random.choice(n_classes, size=n).astype(y_dtype)
+    sample_weight = np.random.random(y.shape).astype(s_w_dtype)
+
+    input_dtypes = (X.dtype, y.dtype, sample_weight.dtype)
+
+    class StrictClassifier(KerasClassifier):
+        def _fit_keras_model(self, X, y, sample_weight, warm_start, **kwargs):
+            assert X.dtype == input_dtypes[0]
+            assert y.dtype == input_dtypes[1]
+            assert sample_weight.dtype == input_dtypes[2]
+            return super()._fit_keras_model(
+                X, y, sample_weight, warm_start, **kwargs
+            )
+
+    clf = StrictClassifier(
+        build_fn=dynamic_classifier, run_eagerly=run_eagerly
+    )
+    clf.fit(X, y, sample_weight=sample_weight)
+    assert clf.score(X, y) >= 0
+    assert clf.predict(X).dtype == y.dtype
+
+
+@pytest.mark.parametrize(
+    "X_dtype", ["float32", "float64", "int64", "int32", "uint8", "uint16"]
+)
+@pytest.mark.parametrize("y_dtype", ["int64", "int32", "uint8", "uint16"])
+@pytest.mark.parametrize("s_w_dtype", ["float32", "float64"])
+@pytest.mark.parametrize("run_eagerly", [True, False])
+def test_regressor_handles_types(X_dtype, y_dtype, s_w_dtype, run_eagerly):
+    """Tests that regressors correctly handle dtype conversions and
+    always return float dtypes.
+    """
+    n, d = 100, 20
+    n_classes = 10
+    X = np.random.uniform(size=(n, d)).astype(X_dtype)
+    y = np.random.choice(n_classes, size=n).astype(y_dtype)
+    sample_weight = np.random.random(y.shape).astype(s_w_dtype)
+
+    input_dtypes = (X.dtype, y.dtype, sample_weight.dtype)
+
+    class StrictRegressor(KerasRegressor):
+        def _fit_keras_model(self, X, y, sample_weight, warm_start, **kwargs):
+            assert X.dtype == input_dtypes[0]
+            assert y.dtype == input_dtypes[1]
+            assert sample_weight.dtype == input_dtypes[2]
+            return super()._fit_keras_model(
+                X, y, sample_weight, warm_start, **kwargs
+            )
+
+    clf = KerasRegressor(build_fn=dynamic_regressor, run_eagerly=run_eagerly)
+    clf.fit(X, y, sample_weight=sample_weight)
+    return_dtype = clf.predict(X).dtype
+    assert return_dtype.kind in ("f", "c")
```

Comment:
Don't we want `assert return_dtype == y.dtype`? Where does complex data enter here?

Reply:
It's expected that if your input y to the regressor was `int32` you will _not_ get `int32` back and instead you'll get a float. I'm 98% sure there would never be a complex float but I included it just in case. If you're 100% sure then by all means make it just `return_dtype.kind == "f"`

Reply:
I just realized that this is just our _test_ and that we do not use complex floats here so... totally safe to remove `"c"`. I also think complex floats are not supported in general (see [here](https://stackoverflow.com/questions/11999147/is-it-possible-to-use-complex-numbers-as-target-labels-in-scikit-learn)).

Reply:
I've made this test stronger in d274a1c7884eac684b873491289c37aba741fd0f, and an accidental casting to `float64`. This relies on the fact that regression problems require floating point outputs.